src/github.com/matrix-org/dendrite/clientapi/writers/joinroom.go

```diff
@@ -0,0 +1,260 @@
+	// A client should only join a room by room ID when it has an invite
+	// to the room. If the server is already in the room then we can
+	// lookup the invite and process the request as a normal state event.
+	// If the server is not in the room the we will need to look up the
+	// remote server the invite came from in order to request a join event
+	// from that server.
+	panic(fmt.Errorf("Joining rooms by ID is not implemented"))
+}
+
+// joinRoomByAlias joins a room using a room alias.
+func (r joinRoomReq) joinRoomByAlias(roomAlias string) util.JSONResponse {
+	domain, err := domainFromID(roomAlias)
+	if err != nil {
+		return util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.BadJSON("Room alias must be in the form '#localpart:domain'"),
+		}
+	}
+	if domain == r.cfg.ServerName {
+		// TODO: Implement joining local room aliases.
+		panic(fmt.Errorf("Joining local room aliases is not implemented"))
+	} else {
+		return r.joinRoomByRemoteAlias(domain, roomAlias)
+	}
+}
+
+func (r joinRoomReq) joinRoomByRemoteAlias(
+	domain gomatrixserverlib.ServerName, roomAlias string,
+) util.JSONResponse {
+	resp, err := r.federation.LookupRoomAlias(domain, roomAlias)
+	if err != nil {
+		switch x := err.(type) {
+		case gomatrix.HTTPError:
+			if x.Code == 404 {
+				return util.JSONResponse{
+					Code: 404,
+					JSON: jsonerror.NotFound("Room alias not found"),
+				}
+			}
+		}
+		return httputil.LogThenError(r.req, err)
+	}
+
+	return r.joinRoomUsingServers(resp.RoomID, resp.Servers)
+}
+
+func (r joinRoomReq) writeToBuilder(eb *gomatrixserverlib.EventBuilder, roomID string) {
+	eb.Type = "m.room.member"
+	eb.SetContent(r.content) // TODO: Set avatar_url / displayname
+	eb.SetUnsigned(struct{}{})
+	eb.Sender = r.userID
+	eb.StateKey = &r.userID
+	eb.RoomID = roomID
+	eb.Redacts = ""
+}
+
+func (r joinRoomReq) joinRoomUsingServers(
+	roomID string, servers []gomatrixserverlib.ServerName,
+) util.JSONResponse {
+	var eb gomatrixserverlib.EventBuilder
+	r.writeToBuilder(&eb, roomID)
+
+	needed, err := gomatrixserverlib.StateNeededForEventBuilder(&eb)
+	if err != nil {
+		return httputil.LogThenError(r.req, err)
+	}
+
+	// Ask the roomserver for information about this room
+	queryReq := api.QueryLatestEventsAndStateRequest{
+		RoomID:       roomID,
+		StateToFetch: needed.Tuples(),
+	}
+	var queryRes api.QueryLatestEventsAndStateResponse
+	if queryErr := r.queryAPI.QueryLatestEventsAndState(&queryReq, &queryRes); queryErr != nil {
+		return httputil.LogThenError(r.req, queryErr)
+	}
+
+	if queryRes.RoomExists {
+		// TODO: Implement joining rooms that already the server is already in.
+		// This should just fall through to the usual event sending code.
+		panic(fmt.Errorf("Joining rooms that the server already in is not implemented"))
+	}
+
+	if len(servers) == 0 {
+		return util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound("No candidate servers found for room"),
+		}
+	}
+
+	var response util.JSONResponse
```

Comment:
Bonus points if you put `var lastErr error` outside the `for` loop and then return that instead of reusing `err`. That would then be clear enough to remove the comments on `:183`.